rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+
+### Single source to multiple sinks
+
+If an event is to be sent to multiple sinks, we want the finalization status to
+only be issued once it has been finalized by all of the sinks. This will
+require the tracking be shared across all the relevant sinks (as opposed to
+being cloned with the event), as well as having a counter or some other
+mechanism for determining when all of the deliveries are completed. The
+topology will need to be involved in setting up this sink counter in order to
+avoid a race condition between handing the event off to the last sink and an
+earlier sink concluding that all deliveries are completed.
+
+### Combining transform (ie `merge`)
+
+If the `merge` transform is configured in a topology, it can produce
+events that are a combination of multiple events merged together. As
+such, their tracking information will also have to be merged, containing
+a reference to all the sources of the events that were combined to produce the
+transformed event. As a complication, this transform may be fed from
+more than one source, resulting in merged events coming from multiple
+sources as well.
+
+### Splitting transform (ie `route`)
+
+This transform has the same kind of effect as does sending a single
+event to multiple sinks, with the distinction that this may send the
+event to one, or several, or even no output pipelines (see below).
+
+### Dropped events (ie `dedupe`, `filter`, or `sample` transforms)
+
+Some transforms optionally drop events depending on a variety of
+conditions. The only realistic action to take at that point is to mark
+the event as having been delivered. It may be useful to maintain
+separate “delivered” vs “dropped” status flags, though this would not
+make a difference to the source acknowledgement.
+
+### User-space transforms (Lua or WASM)
+
+User space transforms may end up doing any of the above behaviours, in
+addition to the possibility of creating entirely new events with no
+relation to the source event. We could potentially automatically tag
+events emitted by a script with the same tracking information as the
+source event, but this may not work correctly if the script is combining
+events. Similar concerns hold for marking events as delivered for
+scripts that do not emit a resulting event. As such, we will have to
+leave this behaviour in the hands of script writers, with Vector
+providing such scripts with convenience library functions to assist with
+the common tasks outlined above.
+
+### Non-acknowledging source
+
+Some sources are unable to provide acknowledgements at a protocol level
+(for example, `apache_metrics`, `prometheus_parser`, or `stdin`).
+Further, users of Vector may wish to selectively enable this feature
+only selectively on certain sources. To minimize the overhead of this
+feature, sources that cannot or should not provide acknowledgements
+should not be required to participate in finalization handling, as they
+would only be discarding the status.
```

Comment:
It seems like we could model this in the docs similar to delivery guarantees.

Reply:
So noted.